weitek osx-project-1 FINAL

weitek opened the pull request.

weitek committed:
start

weitek committed:
FINAL

weitek committed:
add method getByIndex

Добавлен метод для получения доступа к внутреннему массиву по индексу

ohserezha commented on osx-project-1/osx-project-1/main.m:
```diff
@@ -0,0 +1,37 @@
+    @autoreleasepool {
+        
+        id cell1, cell2;
+        srandom((unsigned)time(NULL)); // запуск генератора псевдослучайных чисел
```
перечитал кучу инфы по random, но так и не понял, как он работает. но метод опытного тыка подсказал, что если убрать один запуск генератора, либо в файле NSString+Nucleotides.m изменить random на rand - все нормально генерируется.

weitek replied:
Эти вызовы srandom((unsigned)time(NULL)) перед началом заполнения массива дают такой результат, что у нас получаются идентичные массивы. Если закоментить второй вызов, то массивы будут разными. А одинаковыми я их делал для того чтобы проверить что происходит в предельном случае: вначале оба массива одинаковы, а после мутации на 100% подвергается второй массив.
А вообще говорят что лучше было использовать arc4random.

ohserezha replied:
просто вы закоммитили с одинаковыми, потому я подумал, что вы не нашли решение.
по поводу arc4random - в своем проекте я как раз его и использовал - минимум манипуляций)

ohserezha commented on osx-project-1/osx-project-1/Cell+Mutator.m:
```diff
@@ -0,0 +1,38 @@
+            // выбираем случайную ячейку из списка на замену
+            int randomIndex = (random() % (int)[indexAtDNA count]);
+            // получаем индекс ячейки
+            int randomIndexAtDNA = [[indexAtDNA objectAtIndex:randomIndex] intValue];
```
эта строчка бессмысленна, поскольку randomIndex == randomIndexAtDNA из-за того, что описано выше

weitek replied:
Эта строчка имеет следующий смысл.
После удаления, например второго элемента, массив получается в 0-ой ячейке 0, в 1-ой ячейке 1, а во 2-ой ячейке значение 3.
Этот массив для того чтобы выбирать однозначно те элементы из основного массива которые не прошли процедуру мутации

ohserezha replied:
теперь вижу. когда проверял - поставил после каждой переменной nslog и вывод показывал что они всегда одинаковые, но сейчас перепроверил - так как вы говорите. или магия или я натупил)

weitek replied:
Есть ещё один подобный вариант реализации.
Делается массив в ячейках также лежат индексы.
Выполняют его перемешивание (цикл for в котором случайным образом меняют местами элементы)
А дальше выполняют нужное кол-во замен в основном массиве на основе перемешанного массива с индексами.

ohserezha commented on osx-project-1/osx-project-1/Cell+Mutator.m:
```diff
@@ -0,0 +1,38 @@
+        // список со всеми номерами ячеек которые надо будет менять
+        NSMutableArray *indexAtDNA = [NSMutableArray arrayWithCapacity:DNA_LENGTH];
+        for (int i=0; i<DNA_LENGTH; i++) {
+            [indexAtDNA addObject:[NSNumber numberWithInt:i]];
```
здесь вы заполняете каждую ячейку своим же номером, то есть ячейка 0 хранит 0, 1-1, 2-2 и т.д.

ohserezha commented on osx-project-1/osx-project-1/Cell+Mutator.m:
```diff
@@ -0,0 +1,38 @@
+            // удаляем ячейку из списка замены
+            [indexAtDNA removeObjectAtIndex:randomIndex];
+            // заменяем в выбраной ячейке случайным нуклеотидом
+            [DNA replaceObjectAtIndex:randomIndexAtDNA withObject:[@"" randomNucleotide]];
```
нет проверки на повторные символы. то есть, randomNucleotide может заменить A на А и т.д. В описании проекта это прямо не указано, но символы при замене не должны повторяться.

weitek replied:
Я читал внимательно и несколько раз тех задание для проекта. Там нет четкого указания на этот факт.
Читая комментарии других я видел что из этого в результате выросла проблема! Это говорит о том что ТЗ поставлено пока не профессиональным постановщиком. Да и все это задание - это проба нового инструментария. А алгоритм как убрать повторения можно дописать и я видел несколько очень красивых решений этой проблемы.

ohserezha commented:
Приятно читать так хорошо закомментированный код.

weitek committed:
Change method randomNucleotide

Теперь можно реализовывать мутацию как по оригинальному ТЗ, так и для
проверки что при 100% мутации одной из ДНК - результат мутации будет
100% различие (при условии что исходные ДНК одинаковые)
